### PR TITLE
[gfx90a][conv][nhwc]Asm igemm on gfx90a-fwd/bwd/wrw fp32/fp16

### PR DESCRIPTION
**Pull request by shaojiewang**
Asm igemm implementation-nhwc layout:
||fwd|bwd|wrw|
| ------ | ------ | ------ | ------ |
|fp32| OK | OK | OK |
|fp16| OK | OK | OK |

test script: [test_nhwc](https://amdcloud-my.sharepoint.com/:u:/g/personal/shaowang_amd_com/EZRO3OCUiU1Ks7Y8zT9znJ4BQxSQ2QJHy12mTHc5-fVdSQ?e=8bzcA1)
perf_data: [perf_data](https://amdcloud-my.sharepoint.com/:x:/g/personal/shaowang_amd_com/Ee9OVUVQ4ixJu9s3IX3gjwIBr4nrc44yidtkL4snos4TkA?e=CGDeVV)

### CODE REVIEW

File: src/solver/conv_asm_implicit_gemm_gtc_perf_config.cpp
Comment (carlushuang) [inline src/solver/conv_asm_implicit_gemm_gtc_perf_config.cpp:228]: `kernel_arch` can change to like `gtc`? since here `_gtcx_` or `_gtcx2_` is not exactly arch name

Comment (shaojiewang) [reply src/solver/conv_asm_implicit_gemm_gtc_perf_config.cpp]: yes. changed.

**Comment by carlushuang** [comment]
need wait for https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1031 ready to add ctest for this PR

**Comment by junliume** [comment]
> need wait for #1031 ready to add ctest for this PR

Wait for #1082 to be ready this week (Wk34), then merge develop and check on this one :)

**Comment by junliume** [comment]
@shaojiewang  @carlushuang 
Basic support for gfx90a is merged in develop, could you please modify this branch (if needed) and enable a few tests on gfx90a? Thanks!

**Comment by shaojiewang** [comment]
> @shaojiewang @carlushuang
> Basic support for gfx90a is merged in develop, could you please modify this branch (if needed) and enable a few tests on gfx90a? Thanks!

@junliume Yes!

File: test/CMakeLists.txt
Comment (junliume) [inline test/CMakeLists.txt:531]: This line is causing some issues in CI.

Could we consolidate all "MIOPEN_TEST_GFX90A" to "MIOPEN_TEST_GFX90a", or change existing "MIOPEN_TEST_GFX90a" in develop line to "MIOPEN_TEST_GFX90A"? :)

Comment (shaojiewang) [reply test/CMakeLists.txt:531]: Yes. I think use lowercase to represent ```GFX90a``` is better. 
Modified.

**Comment by junliume** [comment]
@shaojiewang  @carlushuang 
some unit test failures are known issues on gfx90a: https://github.com/ROCmSoftwarePlatform/MIOpen/issues/1093

Do we plan to resolve the above issue in this PR? If not, we can merge develop into this branch, or rebase this branch on top of develop, then the CI should reflect changes in this PR only. Thanks!

**Comment by carlushuang** [comment]
@junliume 
the failure test case in #1093 seems have nonthing to do with asm igemm.
currently the test for asm igemm on gfx90a is `test_conv_igemm_dynamic_xdlops_nhwc_fwd`/`test_conv_igemm_dynamic_xdlops_nhwc_bwd`/`test_conv_igemm_dynamic_xdlops_nhwc_wrw`, and is enabled in this PR.

**Comment by junliume** [comment]
> @junliume
> the failure test case in #1093 seems not directly related to asm igemm. I would say have another PR to enable(or analysis if can be enabled), instead of work in this PR.
> currently the test dedicated for asm igemm on gfx90a is `test_conv_igemm_dynamic_xdlops_nhwc_fwd`/`test_conv_igemm_dynamic_xdlops_nhwc_bwd`/`test_conv_igemm_dynamic_xdlops_nhwc_wrw`, and is enabled in this PR.
> cc @shaojiewang

Yes this approach makes sense and I am on board with it. I've pushed a change to re-disable the unit-test for the time being, and we will work on #1093 after this PR. Thanks!

**Comment by shaojiewang** [comment]
> > @junliume
> > the failure test case in #1093 seems not directly related to asm igemm. I would say have another PR to enable(or analysis if can be enabled), instead of work in this PR.
> > currently the test dedicated for asm igemm on gfx90a is `test_conv_igemm_dynamic_xdlops_nhwc_fwd`/`test_conv_igemm_dynamic_xdlops_nhwc_bwd`/`test_conv_igemm_dynamic_xdlops_nhwc_wrw`, and is enabled in this PR.
> > cc @shaojiewang
> 
> Yes this approach makes sense and I am on board with it. I've pushed a change to re-disable the unit-test for the time being, and we will work on #1093 after this PR. Thanks!

@carlushuang @junliume  I think that the related test is a sub-module of test_conv2d. If we disable test_conv2d, no asm igemm test will run. So it is better to solve the compile error for other asm kernels.

**Comment by junliume** [comment]
> > > @junliume
> > > the failure test case in #1093 seems not directly related to asm igemm. I would say have another PR to enable(or analysis if can be enabled), instead of work in this PR.
> > > currently the test dedicated for asm igemm on gfx90a is `test_conv_igemm_dynamic_xdlops_nhwc_fwd`/`test_conv_igemm_dynamic_xdlops_nhwc_bwd`/`test_conv_igemm_dynamic_xdlops_nhwc_wrw`, and is enabled in this PR.
> > > cc @shaojiewang
> > 
> > 
> > Yes this approach makes sense and I am on board with it. I've pushed a change to re-disable the unit-test for the time being, and we will work on #1093 after this PR. Thanks!
> 
> @carlushuang @junliume I think that the related test is a sub-module of test_conv2d. If we disable test_conv2d, no asm igemm test will run. So it is better to solve the compile error for other asm kernels.

Thanks for the info! We will need to resolve test_conv2d first then. Do we know what might be the cause and how soon to fix please?

**Comment by shaojiewang** [comment]
> We will need to resolve test_conv2d first then. Do we know what might be the cause and how soon to fix please?

The cause is that some algos like winograd or direct conv do not exclude gfx90a arch from their solver. It is not so complicate to fix. I will try to fix by my side. If I do not make it out, I'll find the author to fix it.

**Comment by shaojiewang** [comment]
@junliume @carlushuang @atamazov Please also check and review #1110 for wino solver changes.

**Comment by shaojiewang** [comment]
> BTW how this was tested?

Added in top description.

**Review by junliume — APPROVED** [review]

**Comment by shaojiewang** [comment]
@atamazov Could you please give a second review? Thanks.

**Pull request merged**
